File: app/packages/core/src/session.ts
/**
 * Host side of the LeekWallet session — the counterpart to `src/session.c`.
 *
 * X25519, HKDF-SHA256 into two directional keys, ChaCha20-Poly1305 with counter
 * nonces, and a six-digit passkey. See docs/PROTOCOL.md section 3.
 *
 * The passkey is derived rather than chosen, but that alone was never the
 * mechanism, and v1 of this file claimed it was. Deriving it from the shared
 * secret and nothing else let a relay pick its own key material and search
 * offline for a value reproducing the digits already on the OLED —
 * `sim/passkey_grind.c` did it in 91 seconds on one core. What makes the
 * comparison mean something is the pair of properties this file now
 * implements, taken from BLE Secure Connections numeric comparison (Bluetooth
 * Core Specification v5.4, Vol 3, Part H, §2.3.5.6.4) and ZRTP (RFC 6189
 * §4.4.1.1 and §4.5.2):
 *
 *   - both ends contribute a fresh nonce, and the passkey is bound to the full
 *     transcript — both public keys and both nonces — so substituting any one
 *     of them changes what both screens show;
 *   - the device commits to its nonce before this side reveals its own, so a
 *     relay's every input is fixed before the value that randomises the answer
 *     arrives. It cannot search. It can guess once, online, at 1 in 10^6, in
 *     front of a user who is reading the screen.
 *
 * `verifyCommitment` below is not optional decoration: skipping it puts the
 * offline search straight back, because an uncommitted device nonce is a value
 * a relay may choose after seeing everything else.
 *
 * Uses @noble/*, which viem already depends on — no new supply-chain surface
 * for the one part of this codebase where that would matter most.
 */

import { x25519 } from "@noble/curves/ed25519";
import { hkdf } from "@noble/hashes/hkdf";
import { sha256 } from "@noble/hashes/sha256";
import { chacha20poly1305 } from "@noble/ciphers/chacha";

/* Must match src/session.c exactly. Distinct labels are what keep the two
 * directional keys and the passkey independent outputs of one secret; the v2
 * suffix is what stops a v1 peer and a v2 peer deriving anything usable out of
 * half a handshake if the version check were ever bypassed. */
const LABEL_H2D = "leek-session-h2d-v2";
const LABEL_D2H = "leek-session-d2h-v2";
const LABEL_PASSKEY = "leek-session-passkey-v2";
const LABEL_COMMIT = "leek-session-commit-v2";
const LABEL_TRANSCRIPT = "leek-session-transcript-v2";

/**
 * The wire protocol this client speaks. Must equal PROTOCOL_VERSION in
 * src/protocol.h.
 *
 * v1 had no nonces and no commitment; v2 has both, and the two cannot
 * interoperate. Sent in `hello` and checked in the reply, on both ends, so a
 * mismatched pair says so plainly instead of failing later with "decrypt
 * failed" — which is what v1 did, since it put a version on the wire that
 * nothing ever read.
 */
export const PROTOCOL_VERSION = 2;

/** 128 bits each, as SESSION_NONCE_SIZE in session.h. */
export const NONCE_BYTES = 16;
export const COMMIT_BYTES = 32;

export interface SessionKeys {
  /** Host to device. */
  h2d: Uint8Array;
  /** Device to host. */
  d2h: Uint8Array;
  /** Six digits the user compares against the device screen. */
  passkey: string;
}

/**
 * Everything the derivation is bound to, in the order both ends hash it.
 *
 * By role, never by "mine and theirs": a transcript that depended on who was
 * looking at it would bind nothing, because the two ends would hash different
 * bytes and simply fail to agree.
 */
export interface SessionTranscript {
  hostPublic: Uint8Array;
  devicePublic: Uint8Array;
  hostNonce: Uint8Array;
  deviceNonce: Uint8Array;
}

const enc = new TextEncoder();

function concat(...parts: Uint8Array[]): Uint8Array {
  const out = new Uint8Array(parts.reduce((n, p) => n + p.length, 0));
  let at = 0;
  for (const p of parts) { out.set(p, at); at += p.length; }
  return out;
}

/**
 * The device's commitment to its nonce: H(label ‖ PKb ‖ PKa ‖ Nb).
 *
 * Both public keys are inside the hash for the same reason LESC's f4 takes
 * both: a commitment over the nonce alone could be replayed under a
 * substituted public key.
 */
export function commitment(
  devicePublic: Uint8Array,
  hostPublic: Uint8Array,
  deviceNonce: Uint8Array,
): Uint8Array {
  return sha256(concat(enc.encode(LABEL_COMMIT), devicePublic, hostPublic, deviceNonce));
}

/**
 * Check what the device sent in `helloAck` against what it revealed afterwards.
 *
 * Not constant-time and it does not need to be: both operands are public, and
 * the value being protected is the ORDER of the exchange rather than a secret.
 * Returns false rather than throwing so the caller decides how loudly to fail.
 */
export function verifyCommitment(
  claimed: Uint8Array,
  devicePublic: Uint8Array,
  hostPublic: Uint8Array,
  deviceNonce: Uint8Array,
): boolean {
  const expected = commitment(devicePublic, hostPublic, deviceNonce);
  if (claimed.length !== expected.length) return false;
  let diff = 0;
  for (let i = 0; i < expected.length; i++) diff |= claimed[i]! ^ expected[i]!;
  return diff === 0;
}

/** The whole handshake hashed in a fixed order — the HKDF salt for everything. */
function transcriptHash(t: SessionTranscript): Uint8Array {
  return sha256(concat(
    enc.encode(LABEL_TRANSCRIPT),
    t.hostPublic, t.devicePublic, t.hostNonce, t.deviceNonce,
  ));
}

function derive(shared: Uint8Array, salt: Uint8Array, label: string): Uint8Array {
  return hkdf(sha256, shared, salt, label, 32);
}

/** 128 bits of freshness for this side of the comparison. */
export function generateNonce(): Uint8Array {
  return crypto.getRandomValues(new Uint8Array(NONCE_BYTES));
}

/**
 * Derive session keys and the passkey from a completed X25519 exchange.
 *
 * Throws on a degenerate shared secret: an all-zero result means a small-order
 * peer key, so the "agreement" is a value the attacker chose rather than
 * anything negotiated. Throws too on a transcript of the wrong shape, which is
 * a caller bug rather than a peer one — deriving over a short nonce would
 * silently weaken the very binding this exists for.
 */
export function deriveSession(
  privateKey: Uint8Array,
  peerPublic: Uint8Array,
  transcript: SessionTranscript,
): SessionKeys {
  if (transcript.hostPublic.length !== 32 || transcript.devicePublic.length !== 32 ||
      transcript.hostNonce.length !== NONCE_BYTES ||
      transcript.deviceNonce.length !== NONCE_BYTES) {
    throw new Error("transcript fields are the wrong length");
  }

  const shared = x25519.getSharedSecret(privateKey, peerPublic);

  if (shared.every((b) => b === 0)) {
    throw new Error("degenerate shared secret: peer key is small-order");
  }

  const salt = transcriptHash(transcript);
  const passkeyBytes = derive(shared, salt, LABEL_PASSKEY);
  const n =
    ((passkeyBytes[0]! << 24) >>> 0) +
    (passkeyBytes[1]! << 16) +
    (passkeyBytes[2]! << 8) +
    passkeyBytes[3]!;

  return {
    h2d: derive(shared, salt, LABEL_H2D),
    d2h: derive(shared, salt, LABEL_D2H),
    passkey: String(n % 1000000).padStart(6, "0"),
  };
}

export function generateKeypair(): { privateKey: Uint8Array; publicKey: Uint8Array } {
  const privateKey = x25519.utils.randomPrivateKey();
  return { privateKey, publicKey: x25519.getPublicKey(privateKey) };
}

/** Nonce = 4 zero bytes || 8-byte big-endian counter, as the firmware builds it. */
function nonceFor(counter: number): Uint8Array {
  const nonce = new Uint8Array(12);
  nonce[8] = (counter >>> 24) & 0xff;
  nonce[9] = (counter >>> 16) & 0xff;
  nonce[10] = (counter >>> 8) & 0xff;
  nonce[11] = counter & 0xff;
  return nonce;
}

/**
 * An established session.
 *
 * Counters never reset and never repeat within a session. A reused nonce does
 * not degrade ChaCha20-Poly1305, it breaks it: two messages under one key and
 * nonce leak the keystream and the authentication key.
 */
export type SessionRole = "host" | "device";

export class Session {
  private readonly keys: SessionKeys;
  /* Which key each direction uses depends on which end you are.
   * The host sends under h2d and receives under d2h; the device is the mirror.
   * A single mapping for both ends looks symmetric and silently fails to
   * decrypt, which is how this was found. */
  private readonly sendKey: Uint8Array;
  private readonly recvKey: Uint8Array;
  /** Host only: hold the send counter back until a reply confirms delivery. */
  private readonly deferSend: boolean;
  private txCounter = 0;
  /** What was sealed at the held counter, so a retry can be told from a reuse. */
  private heldPlaintext: Uint8Array | null = null;
  private rxCounter = 0;
  private active = false;

  constructor(keys: SessionKeys, role: SessionRole = "host") {
    this.keys = keys;
    this.sendKey = role === "host" ? keys.h2d : keys.d2h;
    this.recvKey = role === "host" ? keys.d2h : keys.h2d;
    this.deferSend = role === "host";
  }

  get passkey(): string {
    return this.keys.passkey;
  }

  get isActive(): boolean {
    return this.active;
  }

  /**
   * Mark the session usable, once the user has compared the passkey.
   *
   * The host cannot observe the button press, so this is optimistic: it
   * permits encrypted traffic and the device rejects it until the user
   * actually approves. That is safe only because a rejected frame does not
   * advance the send counter - see encrypt().
   */
  confirm(): void {
    this.active = true;
  }

  /**
   * Seal an outgoing frame.
   *
   * The host defers advancing its send counter until a reply proves the frame
   * was accepted. The device advances immediately, mirroring `session_encrypt`
   * in the firmware.
   *
   * The asymmetry exists because only the host retries. It polls an encrypted
   * call while waiting for the user to press ALLOW, and the device rejects
   * every attempt until then without advancing its receive counter. A host
   * that advanced on send would be one ahead before the button was touched,
   * and every later frame would fail to decrypt with nothing to indicate why.
   * Hardware testing found precisely that.
   */
  encrypt(plaintext: Uint8Array): Uint8Array {
    if (!this.active) throw new Error("session not confirmed");

    /* Deferring the counter means the same nonce is used again on a retry, and
     * that is only safe while the plaintext is the same too. Two DIFFERENT
     * messages under one key and nonce is not a degradation of
     * ChaCha20-Poly1305, it is the end of it: the keystreams XOR to the
     * plaintexts and the Poly1305 key falls out, so an eavesdropper reads both
     * messages and can forge a third.
     *
     * Nothing above this class enforces the rule — the client serialises calls
     * and does not poll while it waits for the button, but that is a property
     * of today's caller rather than of the session, and it is invisible to
     * anyone else importing this package. So the rule lives where the nonce
     * does. A retry of the identical request still returns identical bytes,
     * which is what the retry loop relies on; anything else is refused rather
     * than sealed. */
    if (this.deferSend && this.heldPlaintext !== null &&
        !sameBytes(this.heldPlaintext, plaintext)) {
      throw new Error(
        "refusing to seal a second message under one nonce: the previous " +
        "request has not been answered, so the counter has not advanced",
      );
    }

    const out = chacha20poly1305(this.sendKey, nonceFor(this.txCounter)).encrypt(plaintext);
    if (this.deferSend) {
      this.heldPlaintext = plaintext.slice();
    } else {
      this.txCounter++;
    }
    return out;
  }

  /**
   * Open an incoming frame.
   *
   * For the host this is also the acknowledgement that its request was
   * accepted, so the send counter catches up here. For the device only the
   * receive counter moves, as in the firmware.
   */
  decrypt(ciphertext: Uint8Array): Uint8Array {
    if (!this.active) throw new Error("session not confirmed");
    // A failed tag throws. That is correct: a forged frame means the channel is
    // no longer trustworthy, and skipping it and carrying on would be wrong.
    const out = chacha20poly1305(this.recvKey, nonceFor(this.rxCounter)).decrypt(ciphertext);
    this.rxCounter++;
    if (this.deferSend) {
      this.txCounter++;
      /* The held message has been answered, so the nonce it used is spent and
       * the next one is free for anything. */
      this.heldPlaintext = null;
    }
    return out;
  }
}

/**
 * Byte equality. Not constant-time, and it does not need to be: both operands
 * are messages this side composed, so there is no secret here to leak a
 * comparison against.
 */
function sameBytes(a: Uint8Array, b: Uint8Array): boolean {
  if (a.length !== b.length) return false;
  for (let i = 0; i < a.length; i++) {
    if (a[i] !== b[i]) return false;
  }
  return true;
}

File: src/session.c
/**
 * LeekWallet session layer - see session.h
 */

#include "session.h"

#include <string.h>

#include "chacha20poly1305/rfc7539.h"
#include "hmac.h"
#include "sha2.h"
#include "memzero.h"
#include "rand.h"

/* X25519 via the Montgomery ladder directly, rather than
 * curve25519_scalarmult_basepoint(). The "fast basepoint" variant routes
 * through the ed25519 group and its precomputed tables, which is a large
 * dependency for one multiplication and drags Edwards code into a file that
 * only needs Diffie-Hellman. Clamping is done here, as RFC 7748 specifies. */


void curve25519_scalarmult_donna(uint8_t *mypublic, const uint8_t *n,
                                 const uint8_t *basepoint);

static const uint8_t X25519_BASEPOINT[32] = { 9 };

static void x25519_clamp(uint8_t k[32])
{
    k[0]  &= 248;
    k[31] &= 127;
    k[31] |= 64;
}

static void x25519(uint8_t out[32], const uint8_t scalar[32], const uint8_t point[32])
{
    uint8_t clamped[32];
    memcpy(clamped, scalar, 32);
    x25519_clamp(clamped);
    curve25519_scalarmult_donna(out, clamped, point);
    memzero(clamped, sizeof(clamped));
}

/* Domain separators. Distinct labels mean the two directional keys and the
 * passkey are independent outputs of the same shared secret; reusing one label
 * would make them derivable from each other.
 *
 * Bumped to v2 with the commitment handshake. The labels are part of what the
 * two ends agree on, so an old host and new firmware could not accidentally
 * derive a working channel out of half a protocol even if the version check in
 * protocol.c were somehow bypassed — they would simply fail to decrypt. */
static const char LABEL_H2D[]        = "leek-session-h2d-v2";
static const char LABEL_D2H[]        = "leek-session-d2h-v2";
static const char LABEL_PASSKEY[]    = "leek-session-passkey-v2";
static const char LABEL_COMMIT[]     = "leek-session-commit-v2";
static const char LABEL_TRANSCRIPT[] = "leek-session-transcript-v2";

/* ---------------------------------------------------- commitment, transcript */

void session_commitment(const uint8_t device_public[SESSION_PUBKEY_SIZE],
                        const uint8_t host_public[SESSION_PUBKEY_SIZE],
                        const uint8_t device_nonce[SESSION_NONCE_SIZE],
                        uint8_t out[SESSION_COMMIT_SIZE])
{
    /* Plain SHA-256 rather than an HMAC: this is a commitment, not a MAC, and
     * there is no key here for a MAC to take. Binding hygiene comes from the
     * label and from the fixed-width fields, which cannot be re-parsed into a
     * different (key, key, nonce) triple the way a delimiter-free variable
     * layout could. */
    SHA256_CTX ctx;
    sha256_Init(&ctx);
    sha256_Update(&ctx, (const uint8_t *)LABEL_COMMIT, sizeof(LABEL_COMMIT) - 1);
    sha256_Update(&ctx, device_public, SESSION_PUBKEY_SIZE);
    sha256_Update(&ctx, host_public, SESSION_PUBKEY_SIZE);
    sha256_Update(&ctx, device_nonce, SESSION_NONCE_SIZE);
    sha256_Final(&ctx, out);
}

/* The whole handshake, hashed in a fixed order.
 *
 * This is what the keys and the passkey are salted with, and it is why a relay
 * cannot substitute a public key or a nonce on one leg and keep the digits
 * from the other: every one of the four fields is inside the hash, so changing
 * any of them changes what both screens show. Ordered by ROLE — host first,
 * then device — not by who is computing it, or the two ends would hash
 * different bytes and never agree. */
static void session_transcript_hash(const SessionTranscript *t, uint8_t out[32])
{
    SHA256_CTX ctx;
    sha256_Init(&ctx);
    sha256_Update(&ctx, (const uint8_t *)LABEL_TRANSCRIPT, sizeof(LABEL_TRANSCRIPT) - 1);
    sha256_Update(&ctx, t->host_public, SESSION_PUBKEY_SIZE);
    sha256_Update(&ctx, t->device_public, SESSION_PUBKEY_SIZE);
    sha256_Update(&ctx, t->host_nonce, SESSION_NONCE_SIZE);
    sha256_Update(&ctx, t->device_nonce, SESSION_NONCE_SIZE);
    sha256_Final(&ctx, out);
}

/* ------------------------------------------------------------------ HKDF */

/* HKDF-SHA256 (RFC 5869). The salt is the transcript hash rather than a
 * constant: HKDF-Extract's salt is exactly the place for public context that
 * must not be substitutable, and using it means transcript binding costs one
 * SHA-256 rather than a second construction bolted on beside the KDF. */
static void hkdf_sha256(const uint8_t *ikm, size_t ikm_len,
                        const uint8_t salt[32],
                        const char *info, uint8_t out[32])
{
    uint8_t prk[32];
    HMAC_SHA256_CTX ctx;

    /* Extract: PRK = HMAC(salt, IKM) */
    hmac_sha256_Init(&ctx, salt, 32);
    hmac_sha256_Update(&ctx, ikm, (uint32_t)ikm_len);
    hmac_sha256_Final(&ctx, prk);

    /* Expand: one 32-byte block, so T(1) = HMAC(PRK, info || 0x01) */
    size_t info_len = strlen(info);
    const uint8_t counter = 0x01;

    hmac_sha256_Init(&ctx, prk, sizeof(prk));
    hmac_sha256_Update(&ctx, (const uint8_t *)info, (uint32_t)info_len);
    hmac_sha256_Update(&ctx, &counter, 1);
    hmac_sha256_Final(&ctx, out);

    memzero(prk, sizeof(prk));
    memzero(&ctx, sizeof(ctx));
}

/* --------------------------------------------------------------- derivation */

bool session_derive(const uint8_t local_private[SESSION_KEY_SIZE],
                    const uint8_t peer_public[SESSION_PUBKEY_SIZE],
                    const SessionTranscript *transcript,
                    uint8_t k_h2d_out[SESSION_KEY_SIZE],
                    uint8_t k_d2h_out[SESSION_KEY_SIZE],
                    char    passkey_out[SESSION_PASSKEY_LEN + 1])
{
    uint8_t shared[32];
    x25519(shared, local_private, peer_public);

    /* An all-zero shared secret means a small-order peer key: the "agreement"
     * would be a value the attacker chose. Refuse rather than proceed. */
    uint8_t acc = 0;
    for (size_t i = 0; i < sizeof(shared); i++) {
        acc |= shared[i];
    }
    if (acc == 0) {
        memzero(shared, sizeof(shared));
        return false;
    }

    uint8_t salt[32];
    session_transcript_hash(transcript, salt);

    hkdf_sha256(shared, sizeof(shared), salt, LABEL_H2D, k_h2d_out);
    hkdf_sha256(shared, sizeof(shared), salt, LABEL_D2H, k_d2h_out);

    /* Six digits over the shared secret AND the whole transcript.
     *
     * The v1 comment here used to end "a relay in the middle cannot make them
     * agree", and it was wrong: the value was a pure function of the shared
     * secret, so a relay could pick its own key and search offline for one
     * that reproduced the digits already on the device's screen. That search
     * is dead twice over now. The transcript salt means the relay would have
     * to hit a value depending on nonces it does not yet have, and the
     * commitment in session_begin()/session_reveal() means the inputs it does
     * control are fixed before those nonces arrive. What is left is one online
     * guess at 1 in 10^6, in front of a user reading the screen — which is the
     * property BLE LESC's numeric comparison and ZRTP's SAS actually have.
     * See session.h for the citations, and sim/passkey_grind.c for the attack
     * re-run against this construction. */
    uint8_t pk[32];
    hkdf_sha256(shared, sizeof(shared), salt, LABEL_PASSKEY, pk);

    uint32_t n = ((uint32_t)pk[0] << 24) | ((uint32_t)pk[1] << 16) |
                 ((uint32_t)pk[2] << 8) | pk[3];
    n %= 1000000u;
    for (int i = SESSION_PASSKEY_LEN - 1; i >= 0; i--) {
        passkey_out[i] = (char)('0' + (n % 10));
        n /= 10;
    }
    passkey_out[SESSION_PASSKEY_LEN] = '\0';

    memzero(pk, sizeof(pk));
    memzero(shared, sizeof(shared));
    return true;
}

/* ---------------------------------------------------------------- instance */

static struct {
    SessionState state;
    uint8_t      k_h2d[SESSION_KEY_SIZE];
    uint8_t      k_d2h[SESSION_KEY_SIZE];
    char         passkey[SESSION_PASSKEY_LEN + 1];
    uint32_t     rx_counter;
    uint32_t     tx_counter;
    /* Live only between session_begin() and session_reveal(). The private key
     * has to survive the round trip because the shared secret is not computed
     * until the transcript is complete, which is the one cost of the extra
     * leg: a handshake abandoned half-way leaves a private key in RAM until
     * the next session_reset(). Every path that ends a session calls that, and
     * a new hello calls it first thing. */
    SessionTranscript pending_transcript;
    uint8_t      device_private[SESSION_KEY_SIZE];
} sess;

bool session_begin(const uint8_t host_public[SESSION_PUBKEY_SIZE],
                   uint8_t device_public_out[SESSION_PUBKEY_SIZE],
                   uint8_t device_commit_out[SESSION_COMMIT_SIZE])
{
    session_reset();

    /* Ephemeral per connection: a leaked long-term key would otherwise expose
     * every past session. */
    random_buffer(sess.device_private, sizeof(sess.device_private));
    x25519(device_public_out, sess.device_private, X25519_BASEPOINT);

    /* The nonce is generated HERE, before the host's is known, and only its
     * hash goes out. That ordering is the defence: whatever the peer sends
     * next, the device's contribution to the six digits is already fixed and
     * it cannot be revised to land on a chosen value. */
    memcpy(sess.pending_transcript.host_public, host_public, SESSION_PUBKEY_SIZE);
    memcpy(sess.pending_transcript.device_public, device_public_out, SESSION_PUBKEY_SIZE);
    random_buffer(sess.pending_transcript.device_nonce, SESSION_NONCE_SIZE);

    session_commitment(device_public_out, host_public,
                       sess.pending_transcript.device_nonce, device_commit_out);

    /* A small-order host key is still refused, but it can only be caught once
     * the shared secret is computed, and that does not happen until the
     * transcript is complete. So the refusal moved to session_reveal(); the
     * device has revealed nothing but a public key and a hash by then. */
    sess.state = SESSION_AWAITING_REVEAL;
    return true;
}

bool session_reveal(const uint8_t host_nonce[SESSION_NONCE_SIZE],
                    uint8_t device_nonce_out[SESSION_NONCE_SIZE])
{
    /* One reveal per commitment. Answering a second one would hand the peer a
     * fresh derivation against a nonce it had already seen, which is the
     * search this whole round trip exists to prevent. */
    if (sess.state != SESSION_AWAITING_REVEAL) {
        return false;
    }

    memcpy(sess.pending_transcript.host_nonce, host_nonce, SESSION_NONCE_SIZE);

    bool ok = session_derive(sess.device_private,
                             sess.pending_transcript.host_public,
                             &sess.pending_transcript,
                             sess.k_h2d, sess.k_d2h, sess.passkey);
    /* The private key has done its only job. Nothing after this point needs
     * it, so it does not get to sit in RAM for the length of the session. */
    memzero(sess.device_private, sizeof(sess.device_private));

    if (!ok) {
        session_reset();
        return false;
    }

    memcpy(device_nonce_out, sess.pending_transcript.device_nonce, SESSION_NONCE_SIZE);
    sess.state = SESSION_PENDING;
    return true;
}

const char *session_passkey(void)
{
    return sess.passkey;
}

void session_confirm(void)
{
    if (sess.state == SESSION_PENDING) {
        sess.state = SESSION_ACTIVE;
    }
}

/* Whoever cares that the channel died. See session_set_on_reset(). */
static void (*on_reset)(void) = NULL;

void session_set_on_reset(void (*callback)(void))
{
    on_reset = callback;
}

void session_reset(void)
{
    memzero(&sess, sizeof(sess));
    sess.state = SESSION_IDLE;

    /* After the wipe, never before: the callback may look at the session, and
     * it must see one that is already gone. A callback that resets the session
     * itself would recurse, which is why nothing here re-enters. */
    if (on_reset) {
        on_reset();
    }
}

SessionState session_state(void)
{
    return sess.state;
}

/* --------------------------------------------------------------- transport */

/* Nonce = 4 zero bytes || 8-byte big-endian counter. Counters never repeat
 * within a session, and never reset, because a reused nonce with ChaCha20
 * leaks the keystream. */
static void make_nonce(uint32_t counter, uint8_t nonce[12])
{
    memset(nonce, 0, 12);
    nonce[8]  = (uint8_t)(counter >> 24);
    nonce[9]  = (uint8_t)(counter >> 16);
    nonce[10] = (uint8_t)(counter >> 8);
    nonce[11] = (uint8_t)counter;
}

int session_decrypt(uint8_t *data, size_t len)
{
    if (sess.state != SESSION_ACTIVE || len < SESSION_TAG_SIZE) {
        return -1;
    }

    size_t body = len - SESSION_TAG_SIZE;
    uint8_t nonce[12];
    make_nonce(sess.rx_counter, nonce);

    chacha20poly1305_ctx ctx;
    rfc7539_init(&ctx, sess.k_h2d, nonce);
    chacha20poly1305_decrypt(&ctx, data, data, body);

    uint8_t tag[SESSION_TAG_SIZE];
    rfc7539_finish(&ctx, 0, body, tag);
    memzero(&ctx, sizeof(ctx));

    /* Constant-time tag comparison: a byte-at-a-time memcmp would leak how much
     * of a forged tag was correct. */
    uint8_t diff = 0;
    for (size_t i = 0; i < SESSION_TAG_SIZE; i++) {
        diff |= (uint8_t)(tag[i] ^ data[body + i]);
    }
    memzero(tag, sizeof(tag));

    if (diff != 0) {
        /* A failed tag is either corruption or an attack. Either way the
         * session is no longer trustworthy - do not simply skip the frame. */
        session_reset();
        return -1;
    }

    sess.rx_counter++;
    return (int)body;
}

int session_encrypt(uint8_t *data, size_t len, size_t capacity)
{
    if (sess.state != SESSION_ACTIVE || len + SESSION_TAG_SIZE > capacity) {
        return -1;
    }

    uint8_t nonce[12];
    make_nonce(sess.tx_counter, nonce);

    chacha20poly1305_ctx ctx;
    rfc7539_init(&ctx, sess.k_d2h, nonce);
    chacha20poly1305_encrypt(&ctx, data, data, len);
    rfc7539_finish(&ctx, 0, len, data + len);
    memzero(&ctx, sizeof(ctx));

    sess.tx_counter++;
    return (int)(len + SESSION_TAG_SIZE);
}
